### ISSUE TITLE
[no-unsafe-call]: errors on void type

### ISSUE THREAD

**Issue by Hotell** [typescript-eslint/typescript-eslint#2298]
It seems that all `no-unsafe-* ` have this bug - `void` triggers this error. This basically gives one error in all unit tests written with jest, as those matcher contain `void` in return type

**Repro**
```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-call": ["error"]
  }
}
```

```TS
expect(1).toBe(1);
```

![2020-07-15 at 7 31 PM](https://user-images.githubusercontent.com/1223799/87576487-c5f73b00-c6d1-11ea-99e6-5e760edde9d2.png)
![2020-07-15 at 7 32 PM](https://user-images.githubusercontent.com/1223799/87576634-f50dac80-c6d1-11ea-9b27-a5e490f11972.png)

**Expected Result**

no errors for `void`

**Actual Result**

Unsafe call of an any typed value.eslint@typescript-eslint/no-unsafe-call

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.6.1` |
| `@typescript-eslint/parser`        | `3.6.1` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.2` |
| `yarn`                              | `1.22.0` |

**Comment by bradzacher** [comment]
Your example declares a variable `properlyTyped` and then access properties on a variable `nestedAny` - were those names supposed to match?

**Comment by Hotell** [comment]
ah! good catch. looks like you have typos in docs :) I'll send PR if that's ok (for the docs)

- updated the example, to real issue.

TY!

**Comment by Hotell** [comment]
docs fix https://github.com/typescript-eslint/typescript-eslint/pull/2299

**Comment by bradzacher** [comment]
> It seems that all no-unsafe-* have this bug - void triggers this error. 

Note that it's not erroring on the function that's returning `void`, it's erroring on the `toBe` function:

```
   expect(1).toBe(1);
// ^^^^^^^^^^^^^^ unsafe call of an any typed value
```

The rule thinks that `typeof expect(1).toBe === any`. Usually this happens because there's a mis-configuration somewhere meaning that TypeScript can't load the referenced type (i.e. in this case, it hasn't loaded `@types/jest` for some reason).

Would you mind linting that file with `DEBUG=typescript-eslint:eslint-plugin:*`?
I.e. `DEBUG=typescript-eslint:eslint-plugin:* yarn eslint path/to/file.ts`

I'm expecting that you'll see the following log: `Found an "error" any type`.

**Comment by Hotell** [comment]
well, as you said

![2020-07-15 at 9 08 PM](https://user-images.githubusercontent.com/1223799/87585470-699b1800-c6df-11ea-963a-36245220ffe5.png)

**Comment by bradzacher** [comment]
NGL I'm actually kinda surprised that worked exactly 😅 

Okay so now this is where things get difficult. We need to figure out why TS isn't able to find your types.

I'm assuming that there's a separate tsconfig for that (and other) test files? Are you able to dump that?

**Comment by Hotell** [comment]
so I assume you wanna config that is provided within eslint config right ?

<img width="505" alt="image" src="https://user-images.githubusercontent.com/1223799/87590627-62780800-c6e7-11ea-8ea9-06a3c05d5a1f.png">

**Comment by Hotell** [comment]
ha! I found it :)

the culprit is "old good" cypress vs jest global types clash...

tsconfig "include" was including among other things projects that use only jest and also projects that use only cypress (e2e projects/ Im in monorepo environment). So there it is. once I removed those cypress projects  from "include" errors are gone.

wooot thx a ton for leading this investigation man!

**Comment by bradzacher** [comment]
ezpz!

Maybe we should add a debug mode that prints out type resolution errors to help debug this stuff.
It can get pretty complicated to try and resolve this in some more complex setups.

**Comment by Hotell** [comment]
hmm so the "battle" is not over yet it seems 👯 

I run into another one (yes with tests):

```ts
test.each([
    [1, 1, 2],
    [1, 2, 3],
    [2, 1, 3],
  ])('.add(%i, %i)', (a, b, expected) => {
    expect(a + b).toBe(expected);
  });
```

![2020-07-16 at 8 30 AM](https://user-images.githubusercontent.com/1223799/87635013-a4806880-c73e-11ea-8013-493f0a1c41f3.png)

I assume the culprit is the callback typed with `any` as return type ?

![2020-07-16 at 8 31 AM](https://user-images.githubusercontent.com/1223799/87635163-e5787d00-c73e-11ea-8fe6-3cb082a88223.png)

---

What's even more "interesting" that within this function another "bug" is shown within `restrict-plus-operands` rule 

![2020-07-16 at 8 33 AM](https://user-images.githubusercontent.com/1223799/87635305-19ec3900-c73f-11ea-9df7-49fd1aee21d8.png)
![2020-07-16 at 8 35 AM](https://user-images.githubusercontent.com/1223799/87635481-6df71d80-c73f-11ea-8ce5-b3383e9e0d86.png)


- both `a`, `b`, `expected` are inferred to `number`, so there should be no issue. Is this somehow related to how `no-unsafe-*` works?

**Comment by bradzacher** [comment]
in your new example, the problem is likely still your testing library has not had its types loaded - likely they're the same issue.

```ts
   test.each([[1,1,2]])('.add', (a, b, expected) => { const test = a + b });
// ^^^^^^^^^^^^^^^^^^^^ no-unsafe-call
//                                                                 ^^^^^ restrict-plus-operands
```

Stepping through this:
- calling `test.each` does not error. This likely works because the _wrong_ types are loaded, so `test.each` either exists as a non-callable type, or it is `undefined`.
- calling the return value of `test.each()` errors. Following on from the last point - if `test.each` is not typed correctly, then the return type will be an "error any" type, hence this is flagged by `no-unsafe-call`.
- because `test.each()` returns `any`, then when you attempt to call it the arguments all be typed as `any`. This means that the callback you're passing has its arguments `a, b, expected` all typed as `any`.
- Because `a, b` are `any`, this is why they're being flagged by `restricted-plus-operands`.

**Comment by Hotell** [comment]
>  the problem is likely still your testing library has not had its types loaded -

hell yeah it was! this is amazing :D migration to ts-eslint v3 helped me to fix leaking types among whole project.

Thanks a ton @bradzacher . Also big kudos to those new rules API's - they are amazing. If I could I would invite you for a good beer immediately. 

Thanks a ton again.
cheers
